File: User/ugcp_svc/Content/CHIDOutput.cpp
#include "framework.h"
#include "CHIDOutput.h"
#include <thread>

CHIDOutput::CHIDOutput(CProfile& profile, CEventQueue& evQueue, CVirtualHID& vhid)
    : profile(profile), evQueue(evQueue), vhid(vhid)
{
    evExit.set(CreateEvent(NULL, TRUE, FALSE, NULL));
}

CHIDOutput::~CHIDOutput()
{
    SetEvent(evExit.get());
    if (threadClosed.valid())
    {
        WaitForSingleObject(threadClosed.get(), INFINITE);
    }
}

bool CHIDOutput::Init()
{
    output = std::make_unique<CProcessOutput>(profile, vhid);

    threadClosed.set(reinterpret_cast<HANDLE>(_beginthreadex(nullptr, 0, ThreadRead, this, 0, nullptr)));

    return threadClosed.valid();
}

unsigned _stdcall CHIDOutput::ThreadRead(void* param)
{
    auto* local = static_cast<CHIDOutput*>(param);

    SetThreadPriority(GetCurrentThread(), THREAD_PRIORITY_TIME_CRITICAL);

    HANDLE events[3]= { local->evExit.get(), local->evQueue.GetEvQueue(), local->output->GetEvQueue()};
    while (true)
    {

        DWORD ev = WaitForMultipleObjects(3, events, FALSE, INFINITE);
        if ((ev - WAIT_OBJECT_0) != 0)
        {
            CEventPacket* paq = nullptr;
            if ((ev - WAIT_OBJECT_0) == 1)
            {
                paq = local->evQueue.Read();
            }
            local->output->Process(paq);
        }
        else
        {
            break;
        }
    }

    return 0;
}

File: User/ugcp_svc/ProcessOutput/CHIDOutput.cs
using System;

namespace ugcp_svc.ProcessOutput
{
    internal class CHIDOutput : IDisposable
    {
        private readonly Profile.CProfile profile;
        private readonly CVirtualHID vhid;
        private readonly CProcessOutput output;
        private System.Threading.Thread? thread = null;
        private readonly System.Threading.CancellationTokenSource ct = new();

        public CHIDOutput(Profile.CProfile refProfile, CVirtualHID refVhid)
        {
            profile = refProfile;
            vhid = refVhid;
            output = new(profile, vhid);
        }

        void IDisposable.Dispose()
        {
            ct.Cancel();
            thread?.Join();
            ((IDisposable?)output)?.Dispose();
        }

        public bool Init()
        {
            try
            {
                thread = new(() => ThreadRead(ct.Token))
                {
                    Priority = System.Threading.ThreadPriority.Highest
                };
                thread.Start();
            }
            catch
            {
                return false;
            }

            return true;
        }

        private void ThreadRead(System.Threading.CancellationToken token)
        {
            //System.Threading.WaitHandle[] events = [ct.Token.WaitHandle, evQueue.GetEvQueue().WaitHandle, output.GetEvQueue().WaitHandle];

            //while (true)
            //{
            //    int idx = System.Threading.WaitHandle.WaitAny(events);
            //    if (idx != 0)
            //    {
            //        EventQueue.CEventPacket? paq = null;
            //        if (idx == 1)
            //        {
            //            paq = evQueue.Read();
            //        }
            //        output.Process(paq);
            //        return;
            //    }
            //    else
            //    {
            //        return;
            //    }
            //}
        }
    }
}
